src/shogun-core-main/src/main/java/de/terrestris/shoguncore/util/interceptor/secure/WfsRequestInterceptor.java

```diff
@@ -0,0 +1,247 @@
+package de.terrestris.shoguncore.util.interceptor.secure;
+
+import de.terrestris.shoguncore.dao.LayerDao;
+import de.terrestris.shoguncore.dao.UserDao;
+import de.terrestris.shoguncore.model.User;
+import de.terrestris.shoguncore.model.UserGroup;
+import de.terrestris.shoguncore.model.layer.Layer;
+import de.terrestris.shoguncore.model.layer.source.ImageWmsLayerDataSource;
+import de.terrestris.shoguncore.model.layer.source.WfsLayerDataSource;
+import de.terrestris.shoguncore.model.security.Permission;
+import de.terrestris.shoguncore.model.security.PermissionCollection;
+import de.terrestris.shoguncore.service.LayerService;
+import de.terrestris.shoguncore.service.UserService;
+import de.terrestris.shoguncore.util.application.ShogunCoreContextUtil;
+import de.terrestris.shoguncore.util.interceptor.GeoserverAuthHeaderRequest;
+import de.terrestris.shoguncore.util.interceptor.MutableHttpServletRequest;
+import de.terrestris.shoguncore.util.interceptor.WfsRequestInterceptorInterface;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+
+import static org.apache.logging.log4j.LogManager.getLogger;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Interceptor class for WFS requests. Adds basic auth headers based on the GS
+ * properties by default and checks for permission on layer
+ *
+ * @author Nils Bühner
+ * @author Johannes Weskamm
+ *
+ */
+public class WfsRequestInterceptor implements WfsRequestInterceptorInterface {
+
+    /**
+     *
+     */
+    private URI appUri = null;
+
+    /**
+     *
+     */
+    private static final Logger LOG = getLogger(WfsRequestInterceptor.class);
+
+    /**
+     *
+     */
+    @Autowired
+    @Qualifier("userService")
+    protected UserService<User, UserDao<User>> userService;
+
+    /**
+     *
+     */
+    @Autowired
+    @Qualifier("layerService")
+    protected LayerService<Layer, LayerDao<Layer>> layerService;
+
+    /**
+     *
+     */
+    @Value("${geoserver.username:}")
+    private String gsUser;
+
+    /**
+     *
+     */
+    @Value("${geoserver.password:}")
+    private String gsPass;
+
+   /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptGetCapabilities(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS GetCapabilities");
+        return new GeoserverAuthHeaderRequest(request, gsUser, gsPass);
+    }
+
+    /**
+     *
+     */
+    @Override
+    public MutableHttpServletRequest interceptDescribeFeatureType(MutableHttpServletRequest request) {
+        LOG.debug("Intercepting WFS DescribeFeatureType");
+        return new GeoserverAuthHeaderRequest(request, gsUser, gsPass);
```

Comment:
Don't we need to check for permissions here as well?

Reply:
no, as permission is checked in the response interceptor, the request needs to be allowed at first but the layers may be filtered afterwards